### ISSUE TITLE
Question on modifying python script.

### ISSUE THREAD

**Issue by underoath23** [piplongrun/TrailerAddict.bundle#8]
Hello,
I really enjoy this plugin first off. I was wondering if its possible to modify the py script to only grab the trailer options. I dont really want the interviews or behind the scenes stuff. I tried removing the tags in the script for the other extra types and left only the 3 trailer ones in there. However movies that had trailers with all the extras now just have nothing. Wondering if I am doing something wrong or if its just not going to be able to work that way.

Thanks for any help!

**Comment by underoath23** [comment]
I just tried copying in a fresh copy of the repo and am seeing this in the logs.
2017-10-04 17:28:27,401 (2390) :  ERROR (networking:219) - Error opening URL 'https://www.traileraddict.com/ajax/film_popular.php'

**Comment by piplongrun** [comment]
Removing lines [136-150](https://github.com/piplongrun/TrailerAddict.bundle/blob/master/Contents/Code/__init__.py#L136-L150) should do the trick. No need to edit anything else.

I don't know were the "Error opening URL" error comes from. It could be something in the edited code, or maybe the source site is blocking you for doing to many requests(?)

**Comment by underoath23** [comment]
Yea im not sure how I would test that. I just cleaned out the plugins. Rebooted and then readded an unmodified plugin package and get the same message. So maybe it is on the traileraddict side...hmm

**Comment by piplongrun** [comment]
Can you still access the Trailer Addict website from the machine where PMS runs?

**Comment by underoath23** [comment]
Looks like the trailer addict site has blocked my IP 0.o, guessing they have safeguards in place to prevent DDOS. I emailed them explaining what was going on, hoping to be whitelisted. 
Fingers crossed.

**Comment by piplongrun** [comment]
> Looks like the trailer addict site has blocked my IP

I was afraid of that already. I think they have implemented extra protection mechanisms lately, because I also ran into this a while back :|

**Comment by underoath23** [comment]
Here is what traileraddict sent me:

We'll take a look. Cloudflare has updated their protection that looks for "organic" browsing to prevent scraping. The update caused our own mobile apps to start getting blocked until we whitelisted the directories on our end to allow for "scraping" or data collection.

We're not sure what exactly Plex is doing to access TrailerAddict, but it must be making a call outside of normal browser activity, causing the call to be flagged and then later blocked. Maybe it's trying to load the video into its own player instead of the one provided?

We'll ask Cloudflare and get back to you.

**Comment by brunomaso1** [comment]
Im having the same problem, TrailerAddict (or Cloudflare) it's blocking my IP, so im not getting the trailers... @underoath23 Did you have any response about that? Are you still blocked?

**Comment by underoath23** [comment]
@brunomaso1 Havent heard anything back yet. I'll update this thread as soon as I do. I am still currently blocked. The first time it happened the block was temporary. This time seems longer so maybe its truly permanent this time. Hoping they can figure something out.

**Comment by GeneralTweety** [comment]
I have the same problem. TraillerAddict Plugin seems to be the reason why i am blocked on the homepage. I wrote them too and I hope that they will release the bann.

I will inform you if i have further informations.

**Comment by underoath23** [comment]
I sent them a nudge email today and also attached the support doc from cloudflare that says its a setting on their side.
[Cloudflare](https://support.cloudflare.com/hc/en-us/articles/200171886-Error-1006-1007-1008-Access-Denied-Your-IP-address-has-been-banned)

**Comment by piplongrun** [comment]
I also have my ip address blocked :( I'm currently working on putting a caching proxy between the site and my agent to get this going again.

**Comment by piplongrun** [comment]
@underoath23 I think you're going to like this change: https://github.com/piplongrun/TrailerAddict.bundle/commit/b1c444678411e48abb662977cb86f3284360c86b

**Comment by Gershy13** [comment]
So has no one found a fix yet? Ive been blocked too...

**Comment by piplongrun** [comment]
> So has no one found a fix yet? Ive been blocked too...

No, there is no fix. This is a security thing on the side of Cloudflare (DDOS prevention). I am looking into some options to at least fix this agent (but it won't unblock your ip).

**Comment by Gershy13** [comment]
so i have to contact them?

**Comment by piplongrun** [comment]
> so i have to contact them?

You can try, but as you can read above, others have tried that already and do not really hear anything back.

**Comment by Gershy13** [comment]
ahh right... so theres really nothing i can do? Are there any other plugins similar? Or a way to get the trailers from another site and change the plugin?

**Comment by piplongrun** [comment]
I am looking into some options to at least fix this agent.

**Comment by Gershy13** [comment]
so will i be able to use it or is there no hope for me if you do?

**Comment by underoath23** [comment]
you can use it on one condition. You manually refresh meta data slowly. If you do your whole library you will get blocked. If you are already blocked then you are SOL unless trailer addict owners do something

**Comment by Gershy13** [comment]
ahh right, im already blocked, so i highly doubt its gonna get fixed. if only someone can make another plugin that does the same thing with youtube or a different trailer site.

**Comment by piplongrun** [comment]
No need, I hope that my caching proxy will fix the issue as then my own server will take most of the hits instead of Trailer Addict.

**Comment by Gershy13** [comment]
ohh right, but then wouldnt that get banned?

**Comment by underoath23** [comment]
I think he intends to cache the needed meta data server side, so your plugin would query his server for the info instead of always going out to trailer addict for each movie.

**Comment by Gershy13** [comment]
ahh right ok...

**Comment by piplongrun** [comment]
> ohh right, but then wouldnt that get banned?

I hope not :) It will take a while for the cache to build, but we'll see what happens!

Check out release version 2.3: https://github.com/piplongrun/TrailerAddict.bundle/releases/tag/v2.3

**Comment by Gershy13** [comment]
thanks ill give it a shot.

**Comment by Gershy13** [comment]
it still doesnt seem to be working tho? Or have you just not finished it?

**Comment by piplongrun** [comment]
> it still doesnt seem to be working tho? Or have you just not finished it?

Drop in the new version of the agent (v2.3). Restart PMS just to make sure changes are picked up. Go to a movie and try to play a trailer or extra. There should be no need to refresh your library.

**Comment by Gershy13** [comment]
i did all that, and now i cant see any options for extras at all...

**Comment by underoath23** [comment]
Update email from traileraddict I got today:

We only started looking into this today. It's due to the fact we increased our security to medium, which starting blocking requests that came across as scraping. So the Firewall went to work. We are still having scraping issues, so are contemplating increasing the security to high and, once that goes into effect, we are going to see what the Firewall begins blocking and why.

For the time being, we can't clear the firewall in order to see what it builds out and create an algo to block bad requests in the future. Once that is done, we can clear the firewall and let the new security run in its place. But, the hope is that if scraping was taking place the location, proxy or TOR would get banned pretty quick.

**Comment by ewan2395** [comment]
Is the cache site just a reverse proxy to the TrailersAddict site? 

If it isn't, could you possibly share the steps required to creating one.

**Comment by piplongrun** [comment]
> Is the cache site just a reverse proxy to the TrailersAddict site?

Yes, with a fairly long cache time, so we don't keep hammering the source site for the same request over and over and over again.

**Comment by yamisonic** [comment]
I guess I have the same issue since trailers won't play since ~2 weeks.
Here is what I found in plex media server logs at the time corresponding to one of the attempts:
```
Oct 17, 2017 21:39:31.915 [0xa97ffb40] DEBUG - HTTP 500 response from GET http://127.0.0.1:43828/:/plugins/com.plexapp.agents.traileraddict/serviceFunction/url/com.plexapp.agents.traileraddict/Trailer%20Addict/PlayVideo?args=Y2VyZWFsMQoxCnR1cGxlCjAKcjAK&kwargs=Y2VyZWFsMQoxCmRpY3QKMQpzNDEKdGE6Ly90cmFuc2Zvcm1lcnMtdGhlLWxhc3Qta25pZ2h0L3RyYWlsZXJzMwp1cmxyMAo_
Oct 17, 2017 21:39:31.915 [0xa97ffb40] DEBUG - [com.plexapp.agents.traileraddict] HTTP reply status 500, with 3505 bytes of content.
Oct 17, 2017 21:39:31.916 [0xafbfeb40] DEBUG - Completed: [127.0.0.1:48226] 500 GET /:/plugins/com.plexapp.agents.traileraddict/serviceFunction/url/com.plexapp.agents.traileraddict/Trailer%20Addict/PlayVideo?args=Y2VyZWFsMQoxCnR1cGxlCjAKcjAK&kwargs=Y2VyZWFsMQoxCmRpY3QKMQpzNDEKdGE6Ly90cmFuc2Zvcm1lcnMtdGhlLWxhc3Qta25pZ2h0L3RyYWlsZXJzMwp1cmxyMAo_ (10 live) 417ms 3826 bytes
Oct 17, 2017 21:39:31.916 [0xad3fdb40] ERROR - [Transcoder] http://127.0.0.1:32400/:/plugins/com.plexapp.agents.traileraddict/serviceFunction/url/com.plexapp.agents.traileraddict/Trailer%20Addict/PlayVideo?args=Y2VyZWFsMQoxCnR1cGxlCjAKcjAK&kwargs=Y2VyZWFsMQoxCmRpY3QKMQpzNDEKdGE6Ly90cmFuc2Zvcm1lcnMtdGhlLWxhc3Qta25pZ2h0L3RyYWlsZXJzMwp1cmxyMAo_&X-Plex-Token=xxxxxxxxxxxxxxxxxxxx: Server returned 5XX Server Error reply
```
Is this related or pure coincidence?

**Comment by underoath23** [comment]
If you can get to trailersddict website in browser you aren't blocked. If
you can't then that's the issue.

**Comment by yamisonic** [comment]
I was banned but I changed the public IP address of my server which is now able to access traileraddict website.
I might be back in this issue (sic) because I still have no trailer found when testing on Baby Driver, cf. the plugin logs:
```
2017-10-17 22:35:45,900 (-4f9004c0) :  DEBUG (model:234) - Deserializing from /var/lib/plexmediaserver/Library/Application Support/Plex Media Server/Metadata/Movies/9/ebc7a50bc9de39accdd6041aa7b8526966545af.bundle/Contents/com.plexapp.agents.traileraddict/Info.xml
2017-10-17 22:35:45,901 (-4f9004c0) :  DEBUG (networking:166) - Requesting 'http://127.0.0.1:32400/library/metadata/48660/tree'
2017-10-17 22:35:45,913 (-4f9004c0) :  DEBUG (networking:161) - Fetching 'https://api.tadata.me/imdb2ta/v2/?imdb_id=tt3890160' from the HTTP cache
2017-10-17 22:35:45,918 (-4f9004c0) :  DEBUG (networking:166) - Requesting 'https://traileraddict.cache.tadata.me/ajax/film_popular.php'
2017-10-17 22:35:48,104 (-4f9004c0) :  DEBUG (preferences:258) - Loaded preferences from DefaultPrefs.json
2017-10-17 22:35:48,104 (-4f9004c0) :  CRITICAL (agentkit:1078) - Exception in the update function of agent named 'Trailer Addict', called with guid 'com.plexapp.agents.themoviedb://339403?lang=en' (most recent call last):
  File "/usr/lib/plexmediaserver/Resources/Plug-ins-1bf240a65/Framework.bundle/Contents/Resources/Versions/2/Python/Framework/api/agentkit.py", line 1076, in _update
    agent.update(obj, media, lang, **kwargs)
  File "/var/lib/plexmediaserver/Library/Application Support/Plex Media Server/Plug-ins/TrailerAddict.bundle/Contents/Code/__init__.py", line 123, in update
    extra_type = 'deleted_scene'
  File "/usr/lib/plexmediaserver/Resources/Plug-ins-1bf240a65/Framework.bundle/Contents/Resources/Versions/2/Python/Framework/code/sandbox.py", line 108, in <lambda>
    _getitem_           = lambda x, y: x.__getitem__(y),
  File "/usr/lib/plexmediaserver/Resources/Plug-ins-1bf240a65/Framework.bundle/Contents/Resources/Versions/2/Python/Framework/api/runtimekit.py", line 498, in __getitem__
    return self._sandbox.preferences.get()[name]
  File "/usr/lib/plexmediaserver/Resources/Plug-ins-1bf240a65/Framework.bundle/Contents/Resources/Versions/2/Python/Framework/code/preferences.py", line 332, in __getitem__
    raise KeyError("No preference named '%s' found." % name)
KeyError: "No preference named 'add_theatrical_trailers' found."
```
Additional missing info: I use TMDB plugin in my movie library, and enabled TrailerAddict plugin on both TMDB and PlexMovie agents. I configured TrailerAddict agent to get only "Trailers".

**Comment by piplongrun** [comment]
> KeyError: "No preference named 'add_theatrical_trailers' found."

Oops :| That is a typo on my end. Fixed with commit https://github.com/piplongrun/TrailerAddict.bundle/commit/6360099ddbe80f06850f18f870d858616d69ff23

**Comment by piplongrun** [comment]
> If you can get to trailersddict website in browser you aren't blocked. If you can't then that's the issue.

That's not really the case any more since version 2.3 where the caching proxy was introduced into the process. Problem is that certain users have been hammering my proxy so hard (thousands of requests in a short period) that the server is now blacklisted before it could build a decent cache.

**Comment by underoath23** [comment]
Ah I see, well I can verify that for those movies the cache did grab, those are working for me now.

**Comment by piplongrun** [comment]
> Ah I see, well I can verify that for those movies the cache did grab, those are working for me now.

I'm still trying to work around the blocking and the cache is slowly building (only good requests are being cached and matching a movie with a good cache hit will make it work in Plex).

**Comment by MSMortensen** [comment]
Hi,
This plugin is the answer to my problems, great work!
I just installed your plugin and it easily found trailers for all my movies. Only problem is that none of them can be played! 
From the comments here I guess that there might be some problems with being banned from the Traileraddict site. How can I confirm that this is the problem and have you got a solution working?

**Comment by piplongrun** [comment]
> Only problem is that none of them can be played!

Probably due to my proxy server being blocked at that time. I'll re-test as soon as I am back home from work.

> From the comments here I guess that there might be some problems with being banned from the Traileraddict site. How can I confirm that this is the problem and have you got a solution working?

That is correct, heavy scraping gets you an ip ban. But it should not be your home ip any longer now that all traffic flows over the proxy (the proxy will still get banned from time to time).

**Comment by MSMortensen** [comment]
Hi again, 
A bit of an update, some trailers will play (Hatefull 8, the Beguiled) while others still cannot play. I guess that the ones working is the ones you have cached, while the ones not working are due to a IP ban for your proxy that blocks it from getting others.
Will I have to re-scan my library, or will they just start working slowly?

**Comment by piplongrun** [comment]
@MSMortensen As long as the proxy isn't blocked, video playback _should_ work. Hatefull Eight and The Beguiled work indeed (so they're in my cache). Can you give me a couple titles that do not work so I can check if it is a cache problem or maybe something else? Thanks!

**Comment by MSMortensen** [comment]
Well I have just been trying to play trailers from the webinterface instead of from my TV (LG webOS app) and from here the two movies mentioned above doesn't play, but I can play a trailer for "The Haidmaidens Tale", "Jurrasic World", and "Iron Man 3". Among the ones that didn't work from the TV are "Guardians of the Galaxy vol 2", "Kingsman: The Secret Service", "One Flew Over the Cuckoos Nest".

**Comment by piplongrun** [comment]
@MSMortensen Thanks for the examples! I figured out why some trailers did not work and made some changes accordingly (mainly backend changes). Please try again with version [2.6](https://github.com/piplongrun/TrailerAddict.bundle/releases/tag/v2.6) (no need to rescan your library).

**Comment by MSMortensen** [comment]
Fantastic work, it seems that everything is working now!

**Comment by Gershy13** [comment]
Im having another issue after updating.... ive tried restarting my computer, and rescanning and refreshing, but i cant get the trailer button to show up for movies... it works for some tv shows, but not for any movie...

**Comment by piplongrun** [comment]
> i cant get the trailer button to show up for movies

Did you install the latest version of this agant (v2.6 at this time)?

Also, not each and every movie will get trailers from this agent (only movies that are on TA will geet trailers).

> it works for some tv shows, but not for any movie...

This agent is for movies only. Are you sure this agent is activated under _Plex Movie_ and/or _The Movie Database_ and that your library is using one of those agents?

**Comment by piplongrun** [comment]
@Gershy13 Are you maybe the person that has been trying "invalid" URLs on my proxy, for example adding `?page=123` to certain URLs? If so, that probably got you blocked on my proxy server :) Please let me know.

**Comment by Gershy13** [comment]
@piplongrun yes I'm on the latest version, and I was trying multiple movies which are very popular (Spiderman homecoming, the emoji movie, bad boys, etc.) And yes I double checked that it is activated under both, and my library is using Plex movie.
I'm not sure where Plex is getting TV show trailers from then... I don't have any other plugins installed...

And I'm not sure if I'm the one adding URLs, I've just been trying to use the plugin normally?

**Comment by piplongrun** [comment]
@Gershy13 Please upload the agent's log files somewhere so I can have a look. I can't think of anything why it won't work at the moment.

```
com.plexapp.agents.traileraddict.log
com.plexapp.agents.traileraddict.log.1
com.plexapp.agents.traileraddict.log.2
com.plexapp.agents.traileraddict.log.3
com.plexapp.agents.traileraddict.log.4
com.plexapp.agents.traileraddict.log.5
```

Other trailers and TV show trailers probablty come from the default Plex agents. Do you have PlexPass?

**Comment by Gershy13** [comment]
@piplongrun where would i find the log files?

And no i dont have plex pass... im not sure where these trailers are coming from??!

**Comment by piplongrun** [comment]
@Gershy13 Log file locations per OS: https://support.plex.tv/hc/en-us/articles/201106148-Channel-Log-Files

Do you maybe have trailers on disk, next to your movie files?

**Comment by Gershy13** [comment]
@piplongrun no i just checked and there arent any trailers with the movies or tv shows... i cant see the trailer button for any of my movies at all... but for tv shows, i can see it for 2 shows... all the others dont have it... im really confused.

and i just realised ive been putting my plugins in the wrong folder the whole time... trailer addict was installed, but not updated... ill update it and see if it fixes it and ill get back to you.

**Comment by piplongrun** [comment]
> Edit: @piplongrun I just updated the plugin and rescanned, and the movie trailers are back working again yay! The trailers dont look very high quality tho, they look low bitrate 720p.. is this normal? The tv shows still havent changed and are in the same state, im not sure why tho...

Excellent! The channel provides what is available on the source site. I have zero control over that, but yes most stuff is 720p, but hey, everything is available for free :D

And as said, this agent works for movies and not for TV shows. I am going to close this issue now since everything described in here has been solved.
